source/blog/2024-11-05-hanami-220.html.markdown

```diff
@@ -0,0 +1,244 @@
+---
+title: "Hanami 2.2.0: Persistence pays off"
+date: 2024-11-05 12:00:00 UTC
+tags: announcements
+author: Tim Riley
+image: true
+excerpt: >
+  With our new database layer and operations, the Hanami 2 vision is complete!
+---
+
+Two years ago, we [released Hanami 2.0](/blog/2022/11/22/announcing-hanami-200/), opening a new chapter for Hanami and our vision for Ruby apps.
+
+Earlier this year, we took another step and introduced our view layer [with Hanami 2.1](/blog/2024/02/27/hanami-210/).
+
+Today we complete the vision! We are thrilled to share Hanami 2.2 with you. With this release, we introduce a powerful new database layer and a brand new tool for organizing your business logic.
+
+## Persistence pays off: Hanami’s new database layer
+
+Hanami’s new database layer gives you a clear home for all aspects of your database interactions, along with the means to build your own clean interface for your app’s business layer to consume.
+
+When you generate a new app, you’ll have a ready-to-go `DATABASE_URL` in your `.env`. Our default database for new apps is SQLite, with Postgres and MySQL also supported.
+
+You can create a migration with `bundle exec hanami generate migration`, and fill it in:
+
+```ruby
+ROM::SQL.migration do
+  change do
+    create_table :posts do
+      primary_key :id
+      column :title, :text, null: false
+    end
+  end
+end
+```
+
+Now you can migrate your database with `hanami db migrate`.
+
+After this, you can generate a new relation: `hanami generate relation posts`. Relations describe your low-level data sources. Here, this means your database tables. Relations are also your place to add reusable, chainable methods that you can use as the building blocks for expressive, higher-level queries. Add something simple to get started:
+
+```ruby
+module MyApp
+  module Relations
+    class Posts < MyApp::DB::Relation
+      schema :posts, infer: true
+
+      use :pagination
+      per_page 20
+
+      def order_by_latest
+        order(self[:id].desc)
+      end
+    end
+  end
+end
+```
+
+Relations are the primary component of your app’s database layer. While you can interact with them directly, it’s better to build a repo. With repos, you get to build your very own database interface, so you can better manage how your data is accessed across your app. You can build a post repo with `hanami generate repo posts`. Here you can define a method to return your latest posts, built using the methods in your relation:
```

Comment:
Is it true that relations are "primary"? That implies other components are "secondary", etc. I guess so, in the sense that one could get by using only Relations.

Reply:
Good callout, this is confusing language. I've just gone and removed this leading sentence entirely!